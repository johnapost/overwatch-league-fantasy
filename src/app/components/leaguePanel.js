// @flow

import React, { Component } from "react";
import { compose } from "redux";
import { connect } from "react-redux";
import clipboardCopy from "clipboard-copy";
import { List, Icon, Button, Divider, Modal, Input, Form } from "antd";
import uuid from "uuid/v4";
import withFirestore from "../shared/withFirestore";

import type { StoreState } from "../shared/makeStore";
import type { User } from "../shared/user";
import type { League } from "../shared/league";

type Props = {
  firebase: Object,
  firestore: Object,
  id: string,
  // oldInviteLinks: InviteLink[],
  league: League,
  uid: string,
  users: [string, User][] | null
};

type State = {
  creatingLink: boolean,
  latestInviteLink: string | null,
  modalVisible: boolean
};

export class LeaguePanelComponent extends Component<Props, State> {
  state = {
    creatingLink: false,
    latestInviteLink: null,
    modalVisible: false
  };

  createLink = async () => {
    const {
      firebase: { firestore: firestoreDep },
      firestore,
      id,
      uid
    } = this.props;

    this.setState({ creatingLink: true });
    const inviteLinkId = uuid();
    const newInviteLink = {
      leagueId: id,
      createdBy: uid,
      createdAt: firestoreDep.FieldValue.serverTimestamp()
    };

    await firestore.set(
      { collection: "inviteLinks", doc: inviteLinkId },
      newInviteLink
    );

    // Effectively an async setState
    return new Promise(resolve => {
      this.setState(
        {
          creatingLink: false,
          latestInviteLink: inviteLinkId,
          modalVisible: true
        },
        resolve
      );
    });
  };

  openModal = async () => this.setState({ modalVisible: true });

  closeModal = async () => this.setState({ modalVisible: false });

  renderInviteModal = () => {
    const { league } = this.props;
    const { latestInviteLink, modalVisible } = this.state;
    if (!latestInviteLink || typeof window === "undefined") return null;

    const formattedLink = `${
      window.location.origin
    }/?invite=${latestInviteLink}`;
    return (
      <Modal
        title={`Share this invite link to ${league.name}`}
        visible={modalVisible}
        footer={null}
        autoFocusButton={null}
        closable
        onCancel={this.closeModal}
      >
        <Form>
          <Form.Item>
            <Input value={formattedLink} />
          </Form.Item>
          <Button
            name="copy-button"
            type="primary"
            onClick={() => {
              clipboardCopy(formattedLink);
            }}
          >
            Copy link
          </Button>
        </Form>
      </Modal>
    );
  };

  render() {
    const { id, league, uid, users } = this.props;
    const { creatingLink, latestInviteLink } = this.state;

    // TODO: Delete old invite links (createdAt > 1day) in render

    return (
      <>
        {uid === league.ownerUser && (
          <>
            <Button
              name="invite-button"
              type="primary"
              disabled={creatingLink}
              onClick={latestInviteLink ? this.openModal : this.createLink}
            >
              <Icon type={creatingLink ? "loading" : "user-add"} />
              Invite people
            </Button>
            {this.renderInviteModal()}
            {/* TODO: League owner can delete unverified users eventually */}
            <Divider />
          </>
        )}
        {users && (
          <List
            bordered
            dataSource={users}
            itemLayout="vertical"
            renderItem={([userId, user]) => (
              <>
                <List.Item key={`${id}-${userId}`}>
                  <List.Item.Meta
                    title="Team Name"
                    description={`Managed by: ${user.displayName}`}
                  />
                  Roster goes here
                </List.Item>
              </>
            )}
          />
        )}
      </>
    );
  }
}

const filterLeagueUsers = (
  users: {
    [index: string]: User
  },
  id: string
): [string, User][] | null =>
  (id &&
    users &&
    ((Object.entries(users): any): User[]).reduce(
      (accum, [key, value]) =>
        value.userLeagues.includes(id) ? [...accum, [key, value]] : accum,
      []
    )) ||
  null;

export const mapStateToProps = (
  {
    firestore: {
      data: { users }
    },
    user: { uid }
  }: StoreState,
  { id }: Props
) => ({
  users: filterLeagueUsers(users, id),
  uid
});

export default compose(
  connect(mapStateToProps),
  withFirestore(({ id }) => [
    {
      collection: "users",
      where: [["userLeagues", "array-contains", id]]
    }
    // {
    //   collection: "inviteLinks",
    //   where: [["leagueId", "==", id]]
    // }
  ])
)(LeaguePanelComponent);
