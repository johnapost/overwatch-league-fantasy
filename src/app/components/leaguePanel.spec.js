import React from "react";
import { shallow } from "enzyme";
import { LeaguePanelComponent } from "./leaguePanel";

describe("LeaguePanel", () => {
  describe("component", () => {
    const props = {
      firebase: {},
      firestore: {},
      id: "abc123",
      league: {
        drafter: null,
        leagueUsers: ["qwer", "zxcv"],
        name: "Some League",
        ownerUser: "qwer",
        rosterSlots: ["flex", "flex"]
      },
      uid: "zxcv",
      users: [
        ["qwer", { displayName: "owner", userLeagues: ["abc123"] }],
        ["zxcv", { displayName: "user", userLeagues: ["abc123"] }]
      ]
    };

    it("should render fine", () => {
      const wrapper = shallow(<LeaguePanelComponent {...props} />);
      expect(wrapper.html()).toMatchSnapshot();
    });

    describe("when league owner", () => {
      describe("invite button", () => {
        it("should create invite link", () => {});

        it("should open modal if invite link already exists", () => {});
      });

      describe("modal", () => {
        it("should handle copying invite links", () => {});
      });
    });
  });
});
