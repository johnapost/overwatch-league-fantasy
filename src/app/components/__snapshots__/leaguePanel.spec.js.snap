// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaguePanel component should render fine 1`] = `"<div class=\\"ant-divider ant-divider-horizontal\\"></div><div class=\\"ant-list ant-list-vertical ant-list-split ant-list-bordered\\"><div class=\\"ant-spin-nested-loading\\"><div class=\\"ant-spin-container\\"><div class=\\"ant-list-item\\"><div class=\\"ant-list-item-meta\\"><div class=\\"ant-list-item-meta-content\\"><h4 class=\\"ant-list-item-meta-title\\">Team Name</h4><div class=\\"ant-list-item-meta-description\\">Managed by: owner</div></div></div><div class=\\"ant-list-item-content\\">Roster goes here</div></div><div class=\\"ant-list-item\\"><div class=\\"ant-list-item-meta\\"><div class=\\"ant-list-item-meta-content\\"><h4 class=\\"ant-list-item-meta-title\\">Team Name</h4><div class=\\"ant-list-item-meta-description\\">Managed by: user</div></div></div><div class=\\"ant-list-item-content\\">Roster goes here</div></div></div></div></div>"`;
